AjobK/Athena
# 228 reusable popup

closes #228 

<br />

 **What is it supposed to do**  
Adds a reusable component that allows developers to easily implement popups to show errors or other content that requires awareness. Popups are dynamic; the contents can be changed and several actions can be added.
![image](https://user-images.githubusercontent.com/14930277/127669566-2433cb97-ec7a-4707-b817-eb6f99f4d202.png)

<br />

 **How can we test your branch**  
At the moment there is only one scenario where the popup appears. Turn off your local API and try logging in. A popup saying a network error occurred. Try interacting with this popup and see if it is working as you would expect.

Also try playing around with the popup component. The popup component is called from the _loginPrompt_ component (line 70). 
![image](https://user-images.githubusercontent.com/14930277/127671445-2d23504f-f7c1-4aa4-98f9-694c22929e35.png)
To make sure the contents are the same for popups with the same purpose (such as showing a network error), the contents are fetched from _popUpData.js_. 
![image](https://user-images.githubusercontent.com/14930277/127671972-b5e47884-4f89-4afd-8d7d-ac5207552af7.png)
For example, if you want the close button in the popup to not contain text, you can remove the _title_ attribute of the _close_ action object in _actions_ > _close_. The actions themselves (such as closing the popup) are defined in the parent component, so in order to change the actions of a button in a popup, you would have to change them in the parent component (_loginPrompt_).
You could try adding the popup to another component. Just inject the _PopUp_ component with JSX: 
![image](https://user-images.githubusercontent.com/14930277/127672775-e5509742-0203-4629-9dc2-a3d1b054b2b2.png)
Then add a _popup_ attribute to the state of the component (initialise it with _null_). To make the popup appear, change the _popup_ attribute of the state.
![image](https://user-images.githubusercontent.com/14930277/127673109-f6d14980-bbd7-4855-8de9-01d1fc10c1cf.png)
Try to evaluate the process and consider other methods that could be more dynamic, easier to use or cleaner.

<br />

**General checklist**
- [x] Tested happy flow.
- [x] Tested unhappy flow.
- [x] No unexpected exceptions.
- [x] No code with high complexity. (Big O)
- [x] EOF newline.
- [x] No floating comma's and spaces.
- [x] No unexpected/unhandled errors.
- [x] Lines no longer than 'guideline' in IDE.
- [x] Correct indentation.
- [x] No usage of legacy functionalities.
- [x] (JS) Curly braces on the sameline as reserved keywords *(`if () {` instead of `{` on line below)*
- [x] (JS) No double quotes.
- [x] (JS) Single newline above unique code structures and reserved keywords.
- [x] (JS) Variables bundled where possible.
- [x] (JS) No semi-colons except where they have a functional necessity, such as for-loops.

<br />

**If relevant, front-end checklist**
- [x] Responsive. (Desktop all the way to mobile 320px)
- [x] Using SCSS variables where relevant.
- [x] Relevant class names
- [x] No modified global SCSS properties, except if necessary and fully tested the impact.
- [x] Checked [caniuse](https://caniuse.com) for more modern CSS properties.

<br />

**If relevant, back-end checklist**
- [ ] No unhandled promise rejections.
- [ ] Seeders still functional.
- [ ] No primary-key violations.
- [ ] Normalization, atleast [3NF](https://en.wikipedia.org/wiki/Database_normalization).
- [ ] Used guards (middleware) where necessary and tested if route is guarded.

<br />

**If relevant, test these browsers**

---

**PR Rules**  
*1. Add relevant labels.*  
*2. Select atleast two reviewers.*  
*3. Assign all reviewers allowed to merge the branch, including yourself.*  
*4. If relevant to the sprint, put related issue in 'Awaiting Approval' column on project board.*  
*5. Add to the corresponding [milestone](https://github.com/AjobK/Seaqull/milestones).*
